track.py

```diff
@@ -0,0 +1,219 @@
+            angle = math.acos((self.radius**2 + dist**2 - self.end_radius**2)/
+                              (2.0 * self.radius * dist))
+            angle = math.pi/2.0 - angle
+            if self.arc < 0:
+                angle = -angle
+        
+        return math.cos(angle)*dist, math.sin(angle)*dist, self.arc
+
+    def step(self, pose=None):
+        dx, dy, dh = self._step()
+        if pose is None:
+            return dx, dy, dh
+        x, y, heading = pose
+        x += math.cos(heading)*dx - math.sin(heading)*dy
+        y += math.sin(heading)*dx + math.cos(heading)*dy
+        heading += dh
+        return x, y, heading
+
+
+def get_main_track(xmldoc):
+    for section in xmldoc.getElementsByTagName('section'):
+        if section.getAttribute('name') == 'Track Segments':
+            return section
+    return None
+
+
+def get_track_width(xmldoc):
+    for section in xmldoc.getElementsByTagName('section'):
+        if section.getAttribute('name') == 'Main Track':
+            # search for <attnum name="width" unit="m" val="13"/>
+            for attnum in section.getElementsByTagName('attnum'):
+                if attnum.getAttribute('name') == 'width':
+                    return float(attnum.getAttribute('val'))
+    return None
+
+
+def print_track(track):
+    angle = 0.0
+    length = 0.0
+    for segment in track:
+        if segment.arc is not None:
+            angle += segment.arc
+        if segment.length is not None:
+            length += segment.length
+        else:
+            length += segment.arc * segment.radius
+            # TODO variable turns
+        print segment.name
+    print
+    print math.degrees(angle), length
+
+
+def track2xy(track):
+    x, y, heading = (0, 0, 0)
+    minx, miny = 0.0, 0.0
+    for segment in track:
+        x, y, heading = segment.step((x, y, heading))
+        print segment.name, x, y, math.degrees(heading)
+        minx, miny = min(minx, x), min(miny, y)
+    return minx, miny
+
+
+class Track:
+    """Definition of race track"""
+
+    @staticmethod
+    def from_xml_file(filename):
+        xmldoc = parse(filename)
+        segments = []
+        for section in get_main_track(xmldoc).childNodes:
+            if section.nodeType == Node.ELEMENT_NODE:
+                segments.append(Segment.from_xml(section))
+
+        width = get_track_width(xmldoc)
+        return Track(segments, width)
+
+    def __init__(self, segments, width):
+        self.segments = segments
+        self.width = width
+
+    def nearest_segment(self, pose):
+        """Find nearest segment on the track"""
+        global_x, global_y, global_a = pose
+        track_pose = 0, 0, 0
+        for segment in self.segments:
+            # convert global (x,y) into pose relative coordinates
+            x, y, a = track_pose
+            ca, sa = math.cos(-a), math.sin(-a)  # rotate back
+            sx = global_x - x
+            sy = global_y - y
+            sx, sy = ca*sx - sa*sy, sa*sx + ca*sy
```

Comment:
I think it would be great to have these lines in a function with a name, doc string and tests :wink:. I guess that it is trying to convert global car coordinates into segment relative ones.